streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java

```diff
@@ -263,63 +282,44 @@ public boolean isValidTransition(final State newState) {
      * @param listener a new state listener
      */
     public void setStateListener(final KafkaStreams.StateListener listener) {
-        synchronized (stateLock) {
-            if (state == CREATED) {
-                stateListener = listener;
-            } else {
-                throw new IllegalStateException("Can only set StateListener in CREATED state.");
-            }
+        if (state == State.CREATED) {
```

Comment:
In all functions that we only have a single-read on the `state`, we do not need the `stateLock` since it is volatile.

In `setState`, I was originally trying to not keep the user code in the state lock since it may be time consuming. But after thinking about it a bit more, I feel it is safer to keep it in the lock as the listener may be assumed by users to be thread-safe.